File: Level 2/양궁대회/solution.py
from itertools import combinations_with_replacement
from collections import Counter


def calculate_score(combination, info):
    lion_score, apeach_score = 0, 0

    for i in range(1, 11):
        if info[10 - i] < combination.count(i):
            lion_score += i
        elif info[10 - i] > 0:
            apeach_score += i

    return lion_score, apeach_score


def solution(n, info):
    max_diff, max_combination = 0, {}

    for combination in combinations_with_replacement(range(11), n):
        lion_score, apeach_score = calculate_score(combination, info)
        diff = lion_score - apeach_score
        if diff > max_diff:
            max_diff = diff
            max_combination = combination

    if max_diff <= 0:
        return [-1]
    else:
        answer = [0] * 11
        counter = Counter(max_combination)

        for n in counter:
            answer[10 - n] = counter[n]

        return answer

File: Level 2/양궁대회/solution.js
function solution(n, info) {
  let maxDiff = 0;
  let maxCombination;

  combinationsWithRepetition([...Array(11).keys()], n).forEach(
    (combination) => {
      const { lionScore, apeachScore } = calculateScore(combination, info);
      const diff = lionScore - apeachScore;
      if (diff > maxDiff) {
        maxDiff = diff;
        maxCombination = combination;
      }
    }
  );

  if (maxDiff <= 0) {
    return [-1];
  } else {
    const answer = new Array(11).fill(0);
    const scoreCount = maxCombination.reduce((acc, cur) => {
      acc[cur] = (acc[cur] || 0) + 1;
      return acc;
    }, {});
    Object.entries(scoreCount).forEach(([score, count]) => {
      answer[10 - score] = count;
    });
    return answer;
  }
}

function calculateScore(combination, info) {
  let lionScore = 0;
  let apeachScore = 0;

  for (let i = 1; i <= 10; i++) {
    const lionScoreCount = combination.filter((score) => score === i).length;
    if (info[10 - i] < lionScoreCount) {
      lionScore += i;
    } else if (info[10 - i] > 0) {
      apeachScore += i;
    }
  }

  return { lionScore, apeachScore };
}

function combinationsWithRepetition(arr, n) {
  if (n === 1) return arr.map((value) => [value]);

  const result = [];

  arr.forEach((fixed, index, origin) => {
    const rest = origin.slice(index);
    const combinations = combinationsWithRepetition(rest, n - 1);
    const combine = combinations.map((combination) => [fixed, ...combination]);
    result.push(...combine);
  });

  return result;
}
